sympy/core/operations.py

```diff
@@ -504,3 +505,117 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(OpDispatcher):
+    """
+    Handler dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    Two informations need to be registered. The first is the handler, which is registered
+    to a type and returns the result of operation. The second is the priority relation,
+    which is registered to two types and returns the one with higher priority. Order of types
+    does not matter here because both order are searched when dispatching.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, MatAdd, Expr, MatrixExpr
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> add = AssocOpDispatcher("add", Add)
+
+    >>> @add.register_handler(MatrixExpr)
+    ... def _(*args, **kwargs):
+    ...     return MatAdd(*args, **kwargs)
+
+    >>> @add.register_priority(Expr, MatrixExpr)
+    ... def _(expr, mexpr):
+    ...     return mexpr
+
+    >>> from sympy import MatrixSymbol, Symbol
+    >>> a, A = Symbol('a'), MatrixSymbol('A', 2, 2)
+
+    >>> add(a, A, a) == MatAdd(a, A, a)
+    True
+    """
+
+    def __init__(self, name, fallback, doc=None):
+        self.handler_dispatcher = Dispatcher("%s's handler" % name)
+        super().__init__(name, fallback, doc)
+
+    def register_handler(self, cls):
+        """
+        Register the handler function for a type.
+        """
+        def _(func):
+            self.handler_dispatcher.register(cls)(func)
+            return func
+        return _
+
+    def get_handler(self, cls):
+        """
+        Return the registered handler for the argument of given type.
+        """
+        return self.handler_dispatcher.dispatch(cls)
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the argument of higher priority.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
+    def binary_select(self, cls1, cls2):
+        """
+        Select the class using registered priority. If the priority
+        cannot be determined, return ``None``.
+        """
+        result = self.dispatch(cls1, cls2)
+        if result:
+            return result(cls1, cls2)
+        else:
+            result = self.dispatch(cls2, cls1)
+            if result:
+                return result(cls2, cls1)
+            else:
+                return result
+
+    def find_handler(self, args):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for a in args:
+            func, op_priority, handler = type(a), getattr(a, '_op_priority', 0), self.get_handler(type(a))
+            if handler is not None:
+                args_data.append((func, op_priority, handler))
```

Comment:
In the common case all handlers are None and we go through this loop without doing anything. I wonder if that has any noticeable performance impact.  It might be worth profiling to see if this function consumes much time in some calculation that involves a lot of add/mul. I imagine that the time involved is negligible but this is definitely a hot path so it's worth checking. Also when all handlers are None we still compute a bunch of things here that aren't needed. 

Probably the slowest part is the single dispatch. If that has a noticeable impact then it might be worth considering a class attribute instead.

Reply:
Sounds right. I'll try that.

Reply:
Yes that is what I meant but I think what you have here is cleaner. I just suggested that in case profiling shows that this is potentially causing some slowdown.

Reply:
I ran the profiling with `args = [S.One for i in range(1000)]`.

With single dispatch:
```
%%timeit -n 500 -r 10
for a in args:
    _, _, _ = type(a), getattr(a, '_op_priority', 0), add.get_handler(type(a))
```
The result was `3.31 ms ± 18.3 µs per loop (mean ± std. dev. of 10 runs, 500 loops each)`.


With `getattr`:
```
%%timeit -n 500 -r 10
for a in args:
    _, _, _ = type(a), getattr(a, '_op_priority', 0), getattr(a, '_add_handler', None)
```
The result was `143 µs ± 1.76 µs per loop (mean ± std. dev. of 10 runs, 500 loops each)`.

It seem that using `getattr` is about 20 times faster with 1000 args. How do you think? Will using class attribute be better?

Reply:
I think it probably is better to use a class attribute. Note that when I say profile though I mean to time/profile something meaningful like actually computing the `sum`. It might be that this is a tiny unmeasurable effect in actual use.

On the other hand if this is significant then it's probably also worth deferring the call to `type` and `_op_priority` if the `handler` is normally `None`.

Reply:
Migrated to classmethod.